disorder-fsm/src/Disorder/FSM.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LambdaCase #-}
+module Disorder.FSM (
+    Transition(MkTransition)
+  , always
+  , runFSM
+  ) where
+
+import           Prelude (Show(..))
+import           Data.Bool
+import           Data.String
+import           Data.Maybe
+import           Data.Function
+import           Data.Monoid
+import           Control.Monad
+
+import           Test.QuickCheck.Gen
+import           Test.QuickCheck.Monadic
+import           Test.QuickCheck.Property
+
+-- | Defines a transition from state to state
+data Transition e s m = MkTransition {
+    -- | Display name (used in case of failure)
+    name :: String
+    -- | Predicate which determines of this transition is applicable for a given state 's'
+  , preCond :: s -> Bool
+    -- | Transition from state 's' given environment 'e'
+    --   Assertion logic can be placed in 'PropertyM m s'
+  , transition :: e -> s -> PropertyM m s
+  }
+
+-- | Use it as 'preCond' to 'MkTransition' unconditional
+always :: s -> Bool
+always = const True
+
+instance Show (Transition e s m) where
+  show t = name t
```

Comment:
Good question. It is here because `forAll` requires it, but maybe there is a way to make it less restrictive.